### PR TITLE
HARP-7401: Method to retrive map.lookAt() options

### PR DESCRIPTION
**Pull request by pit-rpg**
Signed-off-by: Stepan Krovspei <11714928+pit-rpg@users.noreply.github.com>

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/MapView.ts [hunk 1]
```diff
@@ -2224,7 +2349,7 @@ export class MapView extends THREE.EventDispatcher {
      * in the [[MapViewOptions]]
      */
     renderSync() {
-        this.renderFunc(PerformanceTimer.now());
+        this.render(PerformanceTimer.now());
```
Comment (Astrak) [inline @here/harp-mapview/lib/MapView.ts]: @zbigg @ninok I never understood why we had `renderFunc` there and what it was. Does it serve any purpose?

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 2]
```diff
@@ -1598,7 +1598,8 @@ export class TextElementsRenderer {
         // rendered in screen space. This method is unaffected by near and far clipping planes
         // distances, but may be improved by taking FOV into equation or customizing the
         // focus point screen position based on horizont, actual ground, tilt ets.
-        let factor = this.m_mapView.lookAtDistance / distance;
```
Comment (Astrak) [inline @here/harp-mapview/lib/text/TextElementsRenderer.ts]: This should stay unchanged if `lookAtDistance` is always a number.

File: @here/harp-mapview/lib/MapView.ts [hunk 3]
```diff
@@ -1558,6 +1577,85 @@ export class MapView extends THREE.EventDispatcher {
...
     }
 
+    /**
+     * Returns the camera focus point position in geo-coordinates. If the camera points to an
+     * invalid position will return `null`.
+     */
+    get geoTarget() {
+        const target = MapViewUtils.rayCastWorldCoordinates(this, 0, 0);
+        if (target === null) {
+            throw new Error("Incorrect view properties, the target is null.");
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: Why throwing an error here? This can easily happen if the user is modifying the three.js camera directly

File: @here/harp-mapview/lib/MapView.ts [hunk 4]
```diff
@@ -1558,6 +1577,85 @@ export class MapView extends THREE.EventDispatcher {
...
+        );
+        return THREE.Math.radToDeg(sphericalCoordinates.azimuth + Math.PI);
+    }
+
+    /**
+     * Set `azimuth` property of the [[MapView]] in degrees. Please note [[MapView.geoTarget]] must
+     * be not `null`
+     */
+    set azimuth(azimuthDeg: number) {
+        this.lookAt(this.geoTarget, this.lookAtDistance, this.tilt, azimuthDeg);
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: I think we should log an error if geoTarget is null

File: @here/harp-mapview/lib/MapView.ts (hunk 3 above)
Comment (Astrak) [reply @here/harp-mapview/lib/MapView.ts]: I asked for it :) it's because we have that throughout the code already, and MapView doesn't work if the target is in the void (the method to retrieve the tilt and azimuth doesn't know the target, so lookAt doesn't work)

**Comment by pit-rpg** [comment]
I personally prefer the approach where an error is thrown.
This keeps the code cleaner and also forces the user to use the API created for this, or forces him to do more careful the camera transformations.

File: @here/harp-mapview/lib/MapView.ts [hunk 5]
```diff
@@ -2011,7 +2102,7 @@ export class MapView extends THREE.EventDispatcher {
             // lookAtDistance = (EQUATORIAL_CIRCUMFERENCE * focalLength) / (256 * zoomLevel^2);
             // lookAtDistance = abs(cameraPos.z) / cos(cameraPitch);
             // Here we may use precalculated distance (once pre frame):
-            const lookAtDistance = this.m_lookAtDistance;
+            const lookAtDistance = this.lookAtDistance;
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: Please adjust the comment. It is not precalculated anymore

File: @here/harp-mapview/lib/MapView.ts (hunk 1 above)
Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: looks fine to me to remove it.

File: @here/harp-mapview/lib/MapView.ts (hunk 3 above)
Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: that's not acceptable. We should fix the places where MapView relies on that (I only know that the TiltClipPlaneEvaluator does) and don't add more.
It's totally fine if the map controls have this restriction but MapView should not have it.

**Comment by ninok** [comment]
> I personally prefer the approach where an error is thrown.
> This keeps the code cleaner and also forces the user to use the API created for this, or forces him to do more careful the camera transformations.

There is no way for the user to know if an error will be thrown in this method (we could add it to the documentation but that's easy to miss). Having a GeoCoordinates | undefined return value will make it clear to the user that there is not always a target point.

Also forcing the user to add a try catch is bad b/c it has a performance impact. My rule of thumb is, exceptions should only be used for sth. unexpected. Not having a target point is a totally valid case if the user modified the camera.

File: @here/harp-mapview/lib/MapView.ts [hunk 6]
```diff
@@ -1511,7 +1529,97 @@ export class MapView extends THREE.EventDispatcher {
...
 
-        this.update();
+        this.updateCamera();
+    }
+
+    /**
+     * Returns the camera focus point position in geo-coordinates. If the camera points to an
+     * invalid position will return `null`.
+     */
+    get geoTarget() {
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: I would rather just name this `target`. Did you check the confluence page where we discussed the naming?

File: @here/harp-mapview/lib/MapView.ts [hunk 7]
```diff
@@ -1511,7 +1529,97 @@ export class MapView extends THREE.EventDispatcher {
...
+     * Set `pitch` value of the current view in degrees.
+     */
+    set pitch(pitchDeg: number) {
+        this.setCameraGeolocationAndZoom(this.geoCenter, this.zoomLevel, this.yaw, pitchDeg);
+    }
+
+    /**
+     * Return `yaw` value of the current view in degrees.
+     */
+    get yaw() {
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: why not `azimuth`?

Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: Ah, just saw it below. I would not add yaw and pitch getters/setters at all but we should stick to azimuth/tilt(or however we want to name them)

File: @here/harp-mapview/lib/MapView.ts [hunk 8]
```diff
@@ -1609,24 +1717,24 @@ export class MapView extends THREE.EventDispatcher {
...
+        const zoomLevelDistance = cameraPosZ / Math.cos(Math.min(cameraPitch, Math.PI / 3));
+        return MapViewUtils.calculateZoomLevelFromDistance(zoomLevelDistance, this);
     }
+
     set zoomLevel(zoomLevel: number) {
-        this.m_zoomLevel = THREE.Math.clamp(zoomLevel, this.m_minZoomLevel, this.m_maxZoomLevel);
-        MapViewUtils.zoomOnTargetPosition(this, 0, 0, this.m_zoomLevel);
-        this.update();
+        THREE.Math.clamp(zoomLevel, this.m_minZoomLevel, this.m_maxZoomLevel);
+        this.lookAtDistance = MapViewUtils.calculateDistanceToGroundFromZoomLevel(this, zoomLevel);
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: that's not the inverse of zoomlevel, especially when the pitch is > 60°.

File: @here/harp-mapview/lib/MapView.ts [hunk 9]
```diff
@@ -3294,4 +3332,49 @@ export class MapView extends THREE.EventDispatcher {
...
+            );
+        }
+
+        if (target === undefined) {
+            target = MapViewUtils.rayCastWorldCoordinates(this, 0, 0) || undefined;
+        }
+
+        if (target === undefined) {
+            target = new THREE.Vector3(0.0, 0.0, 1.0);
+            target.applyMatrix4(this.camera.matrixWorld).multiplyScalar(this.m_lookAtDistance);
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: Isn't that camera.getWorldDirection?

File: @here/harp-mapview/lib/MapView.ts [hunk 10]
```diff
@@ -3294,4 +3332,49 @@ export class MapView extends THREE.EventDispatcher {
...
+            const rect = this.canvas.getBoundingClientRect();
+
+            target = this.elevationProvider.rayCast(
+                rect.left + rect.width / 2,
+                rect.top + rect.height / 2
+            );
+        }
+
+        if (target === undefined) {
+            target = MapViewUtils.rayCastWorldCoordinates(this, 0, 0) || undefined;
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: Maybe we should never do the raycasting but always use the formula below. Otherwise it could happen that distance between target and camera is bigger than m_lookAtDistance (especially if tilt > 60°)

File: @here/harp-mapview/lib/MapView.ts [hunk 11]
```diff
@@ -1610,24 +1682,23 @@ export class MapView extends THREE.EventDispatcher {
...
      * Returns the storage level for the given camera setup.
      * Actual storage level of the rendered data also depends on [[DataSource.storageLevelOffset]].
      */
     get storageLevel(): number {
-        return THREE.Math.clamp(
-            Math.floor(this.m_zoomLevel),
-            this.m_minZoomLevel,
-            this.m_maxZoomLevel
-        );
+        return this.zoomLevel;
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: Why did you remove the floor? I think nobody is using this method but we should not break it's beheviour.

File: @here/harp-mapview/lib/MapView.ts [hunk 12]
```diff
@@ -2200,6 +2261,23 @@ export class MapView extends THREE.EventDispatcher {
         this.drawFrame();
     }
 
+    /**
+     * Need to be called if the [[MapView.camera]] position, rotation, quaternion properties are
+     * changed manually.
+     */
+    updateCamera(lookAtDistance?: number) {
+        if (lookAtDistance === undefined) {
+            this.m_lookAtDistance = this.getTarget()
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: Hmm, this is a bit fishy. getTarget will return a vector with length m_lookAtDistance so m_lookAtDistance will effectively never change if the user is moving the camera and no target point is hit.
I think you need to compute the lookAtDistance from the distance to ground (like we do in the zoomlevel getter)

File: @here/harp-mapview/lib/MapView.ts [hunk 13]
```diff
@@ -1610,24 +1682,23 @@ export class MapView extends THREE.EventDispatcher {
      * Returns the zoom level for the given camera setup.
      */
     get zoomLevel(): number {
-        return this.m_zoomLevel;
+        const cameraPitch = MapViewUtils.extractAttitude(this, this.m_camera).pitch;
+        const zoomLevelDistance =
+            this.m_lookAtDistance / Math.cos(Math.min(cameraPitch, Math.PI / 3));
```
Comment (ninok) [inline @here/harp-mapview/lib/MapView.ts]: I think that is wrong. m_lookAtDistance should already be distance to ground divided by cosine of the pitch
